### PR TITLE
[Enhancement] Add default_sort option which allows custom or no default sorting

### PR DESCRIPTION
**Pull request by andrewbridge**
Adds new option `default_sort`, which provides more control over how a full data listing is sorted for display. This is useful for API data which has already been sorted in a specific way for example.

The option recognises either a function or `false`. Where the former will be used to sort the result set when a blank search or a call to `.all()` is made and the latter will disable sorting entirely for full data listings. By default the option is set to the alphabetical search previously seen in the `datastore.sort_results_by` method for backwards compatibility.

This above functionality can also be achieved on a per call basis by calling `store.all(false, false)`.

This PR also includes shims for isArray and fixes old IE inconsistencies with .slice.

This PR also includes an internal datastore function `ds.get_item_key` which moves the standard key getter into its own function so it's reusable during initialisation and a shallowCopy util function.

### Example usage

```
const dataset = [
    {id: 2, item: 'apple', colour: 'red'},
    {id: 8, item: 'banana', colour: 'yellow'},
    {id: 0, item: 'peach', colour: 'orange'},
    {id: 1, item: 'orange', colour: 'orange'},
];

const controlStore = quickspot.datastore({
    data: dataset,
    key_value: 'id'
});

const orderedStore = quickspot.datastore({
    data: dataset,
    key_value: 'id',
    default_sort: false
});

const reverseStore = quickspot.datastore({
    data: dataset,
    key_value: 'id',
    default_sort: (results) => results.sort(
        (a, b) => parseInt(b.__keyvalue) - parseInt(a.__keyvalue)
    )
});

controlStore.store.all();
// controlStore.store.results === [{id: 0, ...}, {id: 1, ...}, {id: 2, ...}, {id: 8, ...}]
controlStore.store.all(false, false);
// controlStore.store.results === [{id: 2, ...}, {id: 8, ...}, {id: 0, ...}, {id: 1, ...}]
orderedStore.store.all();
// orderedStore.store.results === [{id: 2, ...}, {id: 8, ...}, {id: 0, ...}, {id: 1, ...}]
reverseStore.store.all();
// reverseStore.store.results === [{id: 8, ...}, {id: 2, ...}, {id: 1, ...}, {id: 0, ...}]
```

### CODE REVIEW

File: quickspot.js [hunk 1]
```diff
@@ -1238,6 +1252,19 @@
 		};
 
 		/**
+		 * Sort full dataset
+		 * Provides a simple alphabetical sort that should be used when no search has been made.
+		 * While this can provide some odd results for numeric key_values, it is the default for
```
Comment (thybag) [inline quickspot.js]: How so? assuming the keys were numbers, i'd assume it should work as expected?

have tested with;
```
[{"__keyvalue":1},{"__keyvalue":11},{"__keyvalue":21},{"__keyvalue":22}].sort(function(a, b){ return (a.__keyvalue > b.__keyvalue) ? 1 : -1; });
```

**Comment by thybag** [comment]
Can u also bump the npm/bower version etc, given this is goin' master etc :P

File: quickspot.js (hunk 1 above)
Comment (andrewbridge) [reply quickspot.js]: Values are converted to string, so they're sorted lexicographically. Would've been fine for the original use case (i.e. where names are the key value), so it's likely never come up.

```
const dataset = [
    {id: 2, item: 'apple', colour: 'red'},
    {id: 8, item: 'banana', colour: 'yellow'},
    {id: 0, item: 'peach', colour: 'orange'},
    {id: 1, item: 'orange', colour: 'orange'},
    {id: 22, item: 'orange', colour: 'orange'},
    {id: 33, item: 'orange', colour: 'orange'},
    {id: 11, item: 'orange', colour: 'orange'}
];

const controlStore = quickspot.datastore({
    data: dataset,
    key_value: 'id'
});

controlStore.store.all();
// controlStore.store.results === [{id: 0, ...}, {id: 1, ...}, {id: 11, ...}, {id: 2, ...}, {id: 22, ...}, {id: 33, ...}, {id: 8, ...}]
```

**Comment by andrewbridge** [comment]
@thybag All bumped

**Comment by thybag** [comment]
Think u have commited a few extra files

File: quickspot.js (hunk 1 above)
Comment (thybag) [reply quickspot.js]: Ah, yea, forgot about that :/

**Pull request merged**